### ISSUE TITLE
Flops counter issue

### ISSUE THREAD

**Issue by jiagaoxiang** [meta-llama/llama-cookbook#874]
### System Info

Two issues:
1. It seems the current implementation of the flops counter is counting the total flops of all the training step which causes the tflops/s/gpu metric goes up with increasing number of training steps. (reference: https://github.com/pytorch/pytorch/issues/145947#issuecomment-2632532214)
2. actually, it's a question: in the training config,  what is the impact of the default num_freeze_layers: int = 1 on mllama model (llama3.2-11b vision model)? I am suspecting if it is causing the underestimate of total model tflops issue.
 
Thank you!

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### 🐛 Describe the bug

steps to reproduce can be found in this ticket: https://github.com/pytorch/pytorch/issues/145947#issuecomment-2632532214

### Error logs

Explained in above and also in this ticket: https://github.com/pytorch/pytorch/issues/145947#issuecomment-2632532214

### Expected behavior

The Flops counting is buggy.

**Comment by jiagaoxiang** [comment]
@HamidShojanazeri Could you please take a look? Thank you!

**Comment by wukaixingxp** [comment]
@jiagaoxiang  From this [line](https://github.com/meta-llama/llama-cookbook/blob/341539afeeb8c8c50519618c3a7cf8a0a70a6d7f/src/llama_cookbook/finetuning.py#L237), num_freeze_layers will not affect anything if freeze_layers is False (which by default it is [False](https://github.com/meta-llama/llama-cookbook/blob/main/src/llama_cookbook/configs/training.py#L36)).  From your statement `The total flops count with pytorch2.5 and 2.6 is ~186TFlops; but the total flops count with PyTorch nightly (2.7) is ~2900TFlops.`, it looks like an issue with pytorch 2.5 and 2.6, as the issue was gone in pytorch 2.7. It could be Pytorch issue, maybe related to [this FlopCounterMode ignores custom operators issue](https://github.com/pytorch/pytorch/issues/131041)?

**Comment by jiagaoxiang** [comment]
> [@jiagaoxiang](https://github.com/jiagaoxiang) From this [line](https://github.com/meta-llama/llama-cookbook/blob/341539afeeb8c8c50519618c3a7cf8a0a70a6d7f/src/llama_cookbook/finetuning.py#L237), num_freeze_layers will not affect anything if freeze_layers is False (which by default it is [False](https://github.com/meta-llama/llama-cookbook/blob/main/src/llama_cookbook/configs/training.py#L36)). From your statement `The total flops count with pytorch2.5 and 2.6 is ~186TFlops; but the total flops count with PyTorch nightly (2.7) is ~2900TFlops.`, it looks like an issue with pytorch 2.5 and 2.6, as the issue was gone in pytorch 2.7. It could be Pytorch issue, maybe related to [this FlopCounterMode ignores custom operators issue](https://github.com/pytorch/pytorch/issues/131041)?

Hi @wukaixingxp , In my testing, it seems the llama-cookbook flops counter will counter the tflops of all trainnig steps. For example, if I limit the max training step to 1, even if using pytorch nightly, it will give a tflop count of around 190 (for 15 training steps, it will be around 2900Tflops). So, the problem seems to be 2 fold:

1. Llama-cook repo counts flops of all training steps (instead of counting just one step);
2. PyTorch flops counter undercounts the Tflops of a training step; i.e., the total tflops of one training step for the 11B vision model should be much more than 190TFlops with a batch size of 2 and sequence length of 4096.
